### DISCUSSION TITLE
How to request only once when clicking the query button?

### DISCUSSION THREAD

**Discussion by vaynevayne** [TanStack/query#5238] [Q&A]
Expectation: every time an onSubmit is triggered, a query is required, regardless of whether the fetchOptions changes or not

This is my code, and there's a problem that when the form changes, it triggers two queries

What should I do just once? Testers do not accept twice

```jsx
const [form,setForm] = useState()

  const url = '/roiDetail'
  const {
    data: res,
    isFetching,
    error,
    refetch,
  } = useQuery({
    queryKey: [url, fetchOption],
    queryFn: () => axios.post(url, fetchOption),
  })

 <FormContainer
        onSubmit={(newForm: any) => {
          setForm(newForm)
          refetch()
        }}
        form={form}
      />
```

<img width="495" alt="image" src="https://user-images.githubusercontent.com/124666577/230544255-8c7c3924-9b1a-4d22-baf2-35a11e073941.png">

**Comment by TkDodo** [comment]
only call `refetch` if `fetchOptions` haven't changed.

**Reply by vaynevayne** [reply]
Do I need to judge for myself whether fetchOptions has changed? 

Is it OK to go like this? Is there a better way? @TkDodo 
```jsx
 <FormContainer
        onSubmit={(newForm: any) => {
             const hasChanged = JSON.stringify(form) !== JSON.stringify(newForm)
             hasChanged ? setForm(newForm) :  refetch() 
        }}
        form={form}
      />


```

**Reply by TkDodo** [reply]
> Do I need to judge for myself whether fetchOptions has changed?

yes because that's client state.

**Reply by vaynevayne** [reply]
As long as react-query supports custom queryKey comparison functions instead of write-dead serialization, this problem can be solved
